### DISCUSSION TITLE
Update Checker poll

### DISCUSSION THREAD

**Discussion by Matyrobbrt** [neoforged/NeoForge#1079] [Polls]
Hey everyone! We're currently evaluating the merit of the NeoForge-provided update checker. To that end, we're hosting a poll for both mod developers and users to provide feedback on it.  
### **The polls for users can be found [here](https://forms.gle/R6sjPeYWG34SRjw18).**  

Below, you will find a simple poll for mod developers: do you, or do you not include an update checking endpoint in your mods.  
If you have any other thoughts you'd like to share from a developer perspective (like changing the format used), please comment on this discussion. The results of the poll are not binding, though we may act based on its results (improving it, disabling it, removing it, etc.)

**Comment by gingershaped** [comment]
I'll say the same thing here which I did in the user poll: Neo's responsibility is to _load_ mods, and the launcher should handle updating them.

**Reply by robotgryphon** [reply]
+1. Every player I've ever asked is playing some kind of modpack, and they update a whole pack's worth of mods, not individual ones. That'd be the pack software's job (Curseforge, Prism, FTB, etc)

**Reply by TelepathicGrunt** [reply]
The user's poll would have you surprised

**Reply by Shadows-of-Fire** [reply]
Our update checker was introduced to solve the problem of many mods adding their own (often very loud and/or error prone) update checkers.  

Player adoption seems like a poor metric to go by here (I doubt players cared about those either, moreso just annoyed by them). We care about improving UX by reducing the number of poorly written update checkers in the wild.

**Reply by gigaherz** [reply]
Not just error-prone. A ridiculous number of mods had update fetch logic doing http requests in the main thread with default timeout settings, which, when the site hosting the update files inevitably died (temporarily or permanently) caused modpacks to wait for 60+ seconds until the request timed out, sometimes adding multiple minutes to the loading when multiple mods were hosted in the same server.

**Reply by TelepathicGrunt** [reply]
Some modders said they were moving back to their own rolled out update checker because they didn’t like how the forge/neo one works or felt it was too hidden from players. Definitely a problem to keep in mind

**Reply by KnightMiner** [reply]
> Some modders said they were moving back to their own rolled out update checker because they didn’t like how the forge/neo one works or felt it was too hidden from players. Definitely a problem to keep in mind

tbh, I think those modders are just objectively wrong. Their desire for bad UX is not a reason for Neo to drop this feature. There is no way you can in a standalone update checker get something better than all mods having an update checker in a single place. If they think the Neo one has issues, tell them to follow proper open source practices and help improve the Neo one.

**Comment by Gaming32** [comment]
I use the update checker data provided by Modrinth. [Modrinth's docs](https://docs.modrinth.com/#tag/misc/operation/forgeUpdates)

**Reply by gingershaped** [reply]
I could do this if I wanted, but I just don't remember to actually do it :p

**Reply by ApexModder** [reply]
I also rely on Modrinths auto-generated mod version json api

`updateJSONURL = "https://api.modrinth.com/updates/${MOD_ID}/forge_updates.json"`

**Reply by Gaming32** [reply]
Note that you have to append `?neoforge=only` or `?neoforge=include` for NeoForge.

**Reply by gingershaped** [reply]
I could do this if I wanted, but I just don't remember to actually do it :p

**Reply by TelepathicGrunt** [reply]
Curseforge equivalent:
https://curseupdate.com/#documentation:about

**Comment by gigaherz** [comment]
Missing an option: "In some, but it was too tedious to manage for every single one of my mods so I stopped adding them, and then I just stopped updating the other ones too, so for practical purposes no."

**Comment by henkelmax** [comment]
I specifically built a [service](https://github.com/henkelmax/mod-update-server) that automatically keeps track of all my mods versions and generates update JSONs for this.
For example Prisms update checker doesn't work if people download the mod from CurseForge instead of modrinth. This is basically the most reliable way to keep players informed about mod updates. I even made a PR to ModMenu (https://github.com/TerraformersMC/ModMenu/pull/679) so I could add a similar functionality to my Fabric mods. I'd really miss this feature if it were removed from NeoForge. Users always have the option to disable the checker in the config and modders can just not provide an update JSON URL if they don't want to use that feature.

**Reply by SwanX1** [reply]
Same here, I've created a version tracker on [infernalstudios.org](https://infernalstudios.org/api/mods/infernalexpansion/forge)

**Reply by LostLuma** [reply]
On that note, I'd be interested in seeing a way to implement programmatic update checking on NeoForge as well so we can implement the backend the way we'd like to.

**Reply by TelepathicGrunt** [reply]
@LostLuma you mean have the modder be the one making the http call to seeing what version is latest? The point of the update checker being added to the modloader was because modders doing their own checking was doing on the main thread and causing the modpack to stall for minutes at a time if the web request was failing. Modloader’s own checks are done in a safe way to try and stop having modders do their own checks once and for all

**Reply by LostLuma** [reply]
That's what I mean, except that it would still display that updates are available in the unified mod list.
I don't think that it historically having been run on the main thread has to be a reason against it as the mod loader could simply invoke custom update checking logic on separate thread(s) (this is what was done in ModMenu, fwiw).

**Reply by Gaming32** [reply]
Those separate threads could also be virtual threads, which would allow a lot of them to be run together more efficiently (since they're mostly doing web requests).

**Reply by sciwhiz12** [reply]
I personally would be hesitant to cede control over the aspect of the update checker backend to mod-specific implementations, because it might cause fragmentation in the format of the update checker JSON (specified in the `neoforge.mods.toml`) based on whatever the mod fancies writing.

I think having a well-known convention or standard for the update checker JSON (possibly one that transcends NeoForge and works with a variety of loaders), with flexibility and additional information includable, is better for the community as a whole -- mod loaders can implement it, mod distribution platforms can generate it, mod launchers can recognize it, and mod developers can rely on it.

**Comment by Mrbysco** [comment]
Missing an option: "Have too many mods to even look into how to set it up"

**Comment by TeamDman** [comment]
I didn't even realize this was a thing I could have been leveraging. I mainly distribute through CurseForge, Modrinth, and GitHub releases, and expect that anyone using my mods is using something like PrismMC or whatever. When the launcher knows the url of the mod on a distribution platform, I'd then expect that it could then perform updating checking against that provider. I don't know how far prism/others align with these expectations.

That being said, there are scenarios where I'd want to perform an update check from within my mod instead of delegating that responsibility purely to launchers.
![image](https://github.com/neoforged/NeoForge/assets/9356891/c169c522-bb8f-461c-9cc8-c25cbe979e92)

(the "update available" text isn't generated by the mod at this point, and even tho the screenshot come from my 1.19.2 dev env I port all my changes forward using the neo versions that came after)

**Reply by KnightMiner** [reply]
I think for your desire, it be good to be able to query Neo's update check so you can just print the results elsewhere in your mod; if that's not a feature it supports it seems like one it should. Would also mean your logic can respect the update checker being disabled by the modpack dev (since users cannot update in that scenario)

**Comment by moonfather1** [comment]
here are my recommendations:
1) in factory-default mods.toml file, add a line mentioning that modrinth has an api and that the modder doesn't have to maintain a text file manually. i just learned of it today in these comments.
2) in the same file, add an example for curseupdate.com, of which i also learned just now.
3) leave everything else be. some morons think they are so important that their mods should print five lines when the game loads. there is no helping that. come to think of it, blocking chat messages from mods first few seconds wouldn't be a bad idea but i doubt that suggestion would get far.
4) do not change file format. yes, separate files are needed for different loaders, fine. if the system is being under scrutiny, last thing it needs is a format change.
